README.md
# Styled
<a href="https://pub.dev/packages/styled">
  <img src="https://img.shields.io/pub/v/styled.svg" alt="pub.dev"></a>
<a href="https://github.com/rob333/styled/blob/main/LICENSE">
  <img src="https://img.shields.io/github/license/rob333/styled.svg" alt="License"></a>
<a href="https://github.com/rob333/styled/actions">
  <img src="https://github.com/rob333/styled/workflows/Build/badge.svg" alt="Build Status"></a>
<a href="https://github.com/rob333/styled">Styled in Github</a>

<br>

## Features

A declarative UI tool for flutter to simplify the creation of widgets, passint arguments in positional without order.


<table border="0" align="center" cellpadding="0" cellspacing="0">
  <tr>
    <td>
      <img src="https://raw.githubusercontent.com/rob333/styled/main/doc/styled.webp">
    </td>
  </tr>
</table>
<!-- <div style="text-align: center">
<img src="image.jpg" alt="Image description">
</div> -->

<!-- ## Getting started -->

## Use this package as a library

### Depend on it

Run this command:

With Flutter:

```
 $ flutter pub add styled
```

This will add a line like this to your package's pubspec.yaml (and run an implicit flutter pub get):

```yaml
dependencies:
  styled: "^0.0.10"
```

Alternatively, your editor might support flutter pub get. Check the docs for your editor to learn more.


### Import it

Now in your Dart code, you can use:

```dart
import 'package:styled/styled.dart';
```

For help getting started with Flutter, view the online [documentation](https://flutter.dev/docs).

## Usage

- For supported widgets, add a suffix 'd' to the widget name to use the simplified version of it.
- Named arguments precede positional ones.

For example:
```dart
return Scaffoldd(
    AppBard(
      Theme.of(context).colorScheme.inversePrimary,
      Textd(widget.title),
    ),
    Centerd(
      Columnd(
        MainAxisAlignment.center,
        Textd('counter: $_counter'),
        Textd("TextD", 26, 12.5, FontWeight.bold, Colors.red),
        Textd("TextD2", Textd(18.5, FontWeight.w300, Colors.orange)),
        Rowd(
          MainAxisAlignment.center,
          Containerd(50, 50, const Color.fromARGB(255, 255, 0, 0)),
          Containerd(50, 50, const Color.fromARGB(255, 231, 255, 10)),
          Containerd(50, 50, const Color.fromARGB(255, 15, 221, 15)),
        ),
        Rowd(
          MainAxisAlignment.center,
          Containerd(35, 35, Colors.red, BoxShape.circle, Borderd(3, Colors.redAccent)),
          Containerd(Colors.yellow, BoxShape.circle, 35, 35, Borderd(3, Colors.yellowAccent)),
          Containerd(35, 35, Borderd(3, Colors.greenAccent), Colors.green, BoxShape.circle),
        ),
      ),
    ),
    FloatingActionButtond(
        _incrementCounter,
        'Increment',
        const Icon(Icons.add),
    ),
  );
```

Explained version:
```dart
return Scaffoldd(
    AppBard( // `AppBar` => 'appBar' of Scaffold
      Theme.of(context).colorScheme.inversePrimary, // `Color:0` => 'backgroundColor'
      Text(widget.title), // `Widget` => 'title'
      title: Textd("named arguments precede positional ones", 20,
          const Color.fromARGB(255, 33, 211, 71)), // override 'title' with named argument
      ),
    ),
    Centerd( // `Widget` => 'body' of Scaffold
      Columnd( // `Widget` => 'child' of Center
        MainAxisAlignment.center,
        Textd('counter: $_counter'), // `Widget` will be added into the children of Column|Row.
        // The inner Textd returns a `TextStyle`, when no text:String is specified.
        Textd("TextD2", Textd(18.5, FontWeight.w300, Colors.orange)),
        <Widget>[ // type of `List<Widget>` => 'children' of the Column
          Textd(
            "TextD", // `String` => text to display
            26,   // 1st of type `int|double` => 'fontSize'
            12.5, // 2nd of type `int|double` => 'letterSpacing'
            FontWeight.bold,
            FontStyle.italic,
            // 1st of type `Color` => 'color'
            Colors.red,
            // 2nd of type `Color` => 'backgroundColor'
            const Color.fromARGB(255, 55, 101, 228),
            // 3rd of type `Color` => 'decorationColor'
            const Color.fromARGB(255, 226, 223, 33),
            // use enum:TextMaxLines to set maxlines
            TextMaxLines.three,
            // override `backgroundColor` with named argument
            backgroundColor: Colors.greenAccent,
          ),
        ],
      ),
    ),
    FloatingActionButtond( // => 'floatingActionButton' of Scaffold
        _incrementCounter,     // `void Function()` => 'onPressed' 
        'Increment',           // `String` => 'tooltip'
        const Icon(Icons.add), // `Widget` => 'child'
    ),
  );
```


## Additional information

- `Textd`: Return a Text widget or a TextStyle if no String type argument is specified. 

- In the hint document, for example, `Color: 0:color, 1:backgroundColor, 2: decorationColor, 3:selectionColor` means the first Color type argument is treated as 'color', the second is 'backgroundColor', the third is 'decorationColor', and the fourth is 'selectionColor'.

- `Containerd` mixin with the `Decoration` arguments.
  - Precedence of mixin arguments, `Decoration` in the `Containerd`.
    1. Positional arguments of `Decoration` (Lowest)
    2. Positional `Decoration` object.
    3. Named arguments of `Decoration`
    4. Named `Decoration` object. (Highest)
   
For example:

With ***BoxDecoration***:
```dart
  Containerd(
    45, // width
    45, // height
    BoxDecorationd(
      Colors.blue,
      Borderd(Colors.black, 3),
      BoxShadowd(Colors.grey.withOpacity(0.5), 7, 5, const Offset(0, 3)),
      BorderRadius.circular(0.5),
      borderRadius: BorderRadius.circular(10.5), // named arg precedes positional one
    ),
  ),
```
With ***Mixin***:
```dart
  Containerd(
    45, // width
    45, // height
    Colors.blue,
    Borderd(Colors.black, 3),
    BoxShadowd(Colors.grey.withOpacity(0.5), 7, 5, const Offset(0, 3)),
    BorderRadius.circular(0.5),
    borderRadius: BorderRadius.circular(10.5), // named arg precedes positional one
  ),
```


## Supported Widgets

<table><tr>
<td>
<ul> <!--style="column-count:5"-->
<li> Textd
<li> Centerd
<li> Rowd
<li> Columnd
<li> Scaffoldd (partial)
<li> AppBard (partial)
<li> FloatingActionButtond (partial)
<li> Drawerd
<li> SafeAread
<li> Containerd
<li> Paddingd
<li> Stackd
<li> Positionedd
<li> ElevatedButtond
<li> SizedBoxd
<li> LimitedBoxd
<li> Spacerd
<li> ListViewd (partial)
<li> ListTiled (partial)
<li> GridViewd (partial)
<li> GridTiled
<li> Expandedd
<li> Wrapd
<li> Opacityd
<li> PageViewd (partial)
<li> Tabled
<li> TableRowd
<li> SliverAppBard (partial)
<li> ClipRRectd
<li> ClipOvald
<li> ClipPathd
<li> ClipRectd
<li> Herod
<li> CustomPaintd
<li> ToolTipd
<li> IconButtond
<li> FittedBoxd
<li> AbsorbPointerd
<li> Rotated as Transform.rotate
<li> Scaled as Transform.scale
<li> Translated as Transform.translate
<li> Flipd as Transform.flip
<li> BackdropFilterd
<li> Alignd
<li> Dismissibled
<li> Flexibled
<li> AspectRatiod
<li> Placeholderd
<li> RichTextd
<li> IndexedStackd
</ul>
</td><td>
<ul>
<li> AnimatedContainerd
<li> AnimatedBuilder
<li> AnimatedListd
<li> AnimatedIcond
<li> AnimatedSwitcherd
<li> AnimatedPositionedd
<li> AnimatedPaddingd
<li> AnimatedOpacityd
<li> AnimatedCrossFaded
<li> Semanticsd (partial)
<li> ConstrainedBoxd
<li> FractionallySizedBoxd
<li> SingleChildScrollViewd
<li> DataTabled (partial)
<li> Sliderd
<li> SliderAdaptived as Slider.adaptive
<li> RangeSliderd
<li> AlertDialogd
<li> AlertDialogAdaptived as AlertDialog.adaptive
<li> DraggableScrollableSheetd
<li> ColorFilteredd
<li> ToggleButtonsd
<li> Imaged
<li> ImageAssetd
<li> ImageFiled
<li> ImageMemoryd
<li> ImageNetworkd
<li> DefaultTabControllerd
<li> TabBard
<li> Tabd
<li> TabBarViewd
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
</ul>
</td><td>
<ul>
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
</ul>
</td>
</tr></table>


### 
<ul style="column-count:5">
<li> BoxShadowd
<li> BoxDecorationd
<li> LinearGradientd
<li> BorderSided
<li> Borderd
<li> DecorationImaged
<li> ImageBlurd as ImageFilter.blur
<li> ImageDilated as ImageFilter.dilate
<li> ImageEroded as ImageFilter.erode
<li> ImageCompose as ImageFilter.compose
<li> TextSpand
<li> DataRowd
<li> DataColumnd
<li> DataCelld
<!-- <li>
<li> 
<li>
<li>
<li>
<li>
<li>
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li> 
<li>  -->
</ul>

### Known not supported:
<ul>
<li> layoutBuilder
<li> Positioned.*
<li> ValueListenableBuilder 
<li> ReorderableListView
<li> ReorderableListView.builder
<li> SelectableText
<li> SelectableText.rich
<li> ---
<li> TableBorder
<li> ImageFilter.matrix
</ul>


## Changelog

Please see the [Changelog](https://github.com/rob333/styled/blob/main/CHANGELOG.md) page.



## License

Styled is distributed under the MIT License. See [LICENSE](https://github.com/rob333/styled/blob/main/LICENSE) for more information.


CHANGELOG.md
## 0.0.10

* Mainly adds `Containerd` with mixin `Decoration` arguments.

## 0.0.9

* Initial release.


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:styled/styled.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  static List<Tab> myTabs = [
    Tabd('LEFT', const Icon(Icons.join_left)),
    Tabd('STRAIGHT', const Icon(Icons.straight)),
    Tabd('RIGHT', const Icon(Icons.join_right)),
  ];

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: myTabs.length,
      child: Scaffoldd(
        AppBard(
          Theme.of(context).colorScheme.inversePrimary, // backgroundColor
          Text(widget.title), // title
          title: Textd("named arguments precede positional ones", 20, const Color.fromARGB(255, 33, 211, 71)),
          TabBard(myTabs),
        ),
        Rowd(
          MainAxisAlignment.center,
          Centerd(
            Columnd(
              MainAxisAlignment.center,
              Textd(
                'counter: $_counter',
                backgroundColor: Colors.amberAccent,
                style: Theme.of(context).textTheme.headlineMedium,
              ),
              Textd("text TextStyle", Theme.of(context).textTheme.bodyLarge),
              Textd("Styled TextD", Textd(18.5, FontWeight.w300, Colors.orange)),
              Textd("Styled TextD2", 18.5, backgroundColor: Colors.amber),
              Textd(
                "TextD", // text
                26, // 1st of type `int|double` which is 'fontSize'
                12.5, // 2nd of type `int|double` which is 'letterSpacing'
                FontWeight.bold,
                FontStyle.italic,
                // 1st of type `Color`, which is 'color'
                Colors.red,
                // 2nd of type `Color`, which is 'backgroundColor'
                const Color.fromARGB(255, 55, 101, 228),
                // 3rd of type `Color`, which is 'decorationColor'
                const Color.fromARGB(255, 226, 223, 33),
                // use enum:TextMaxLines to set maxlines
                TextMaxLines.three,
                // override `backgroundColor` with named argument
                backgroundColor: Colors.greenAccent,
              ),
              Rowd(
                MainAxisAlignment.center,
                Containerd(35, 35, const Color.fromARGB(255, 255, 0, 0)),
                Containerd(35, 35, const Color.fromARGB(255, 231, 255, 10)),
                Containerd(35, 35, const Color.fromARGB(255, 15, 221, 15)),
              ),
              Rowd(
                MainAxisAlignment.center,
                Containerd(35, 35, Colors.red, BoxShape.circle, Borderd(3, Colors.redAccent)),
                Containerd(Colors.yellow, BoxShape.circle, 35, 35, Borderd(3, Colors.yellowAccent)),
                Containerd(35, 35, Borderd(3, Colors.greenAccent), Colors.green, BoxShape.circle),
              ),
              Containerd(
                Paddingd(
                  const EdgeInsets.all(5.5),
                  Textd('This is a fancy container!', Colors.blue),
                ),
                LinearGradientd(Colors.red, Colors.orange, Colors.yellow),
                BorderRadius.circular(20.0),
                BoxShadowd(
                  Colors.black.withOpacity(0.5),
                  10,
                  const Offset(0.0, 5.0),
                ),
              ),
              Containerd(
                45,
                45,
                Colors.blue,
                Borderd(Colors.black, 3),
                BorderRadius.circular(0.5),
                BoxShadowd(
                  Colors.grey.withOpacity(0.5),
                  7,
                  5,
                  const Offset(0, 3),
                ),
                borderRadius: BorderRadius.circular(10.5), // named arg precedes
                // named decoration precedes
                decoration: BoxDecorationd(
                  LinearGradientd(
                    Colors.red,
                    Colors.orange,
                    Colors.yellow,
                    Alignment.topLeft,
                    Alignment.bottomRight,
                  ),
                  BorderRadius.circular(20.0),
                  BoxShadowd(
                    Colors.black.withOpacity(0.5),
                    10,
                    const Offset(10.0, 10.0),
                  ),
                ),
              ),
              RichTextd(
                TextSpand(
                  'This\n',
                  Textd(25, Colors.red, FontWeight.bold),
                  TextSpand('is link\n', Textd(20, Colors.blue, TextDecoration.underline)),
                  TextSpand('text.\n', Textd(Colors.yellow, FontStyle.italic)),
                  TextSpand('bold\n', Textd(Colors.purple, FontWeight.bold)),
                ),
              ), // RichTextd
            ),
          ),
          Columnd(
            MainAxisAlignment.center,
            Stackd(
              Containerd(80, 80, Colors.blue),
              Containerd(60, 60, Colors.yellow),
              Containerd(50, 50, Colors.red),
            ),
            Stackd(
              Alignment.center,
              Containerd(80, 80, Colors.blue, BoxShape.circle),
              Containerd(60, 60, Colors.yellow, BoxShape.circle),
              LimitedBoxd(30, 30, Containerd(50, 50, Colors.red, BoxShape.circle)),
            ),
            Stackd(
              Containerd(50, 50, color: Colors.white),
              Containerd(
                const EdgeInsets.all(5.0),
                Alignment.bottomCenter,
                LinearGradientd(
                  Alignment.topCenter,
                  Alignment.bottomCenter,
                  Colors.black.withAlpha(0),
                  Colors.black12,
                  Colors.black45,
                ),
                Textd('Foreground', Colors.white, 20),
              ),
            ),
            ElevatedButtond(
              Textd("Button", 16, Colors.blue), // `child`
              () {}, // `onPressed`
            ),
            // BackdropFilterd(
            //   ImageBlurd(5, 6), //ImageFilter.blur(sigmaX: 5, sigmaY: 6),
            ElevatedButtond(
              const Icon(Icons.ac_unit_rounded), // `icon`
              Textd("label", 16), // `label`
              () {}, // `onPressed`
            ),
            // ),
            Scaled(
              0.8,
              1.2,
              IconButtond(
                26.5, // `iconSize`
                8.5, // `splashRadius`
                const Icon(Icons.point_of_sale_rounded),
                "tooltip", // `tooltip`
                () {}, // `onPressed`
                Colors.green,
                Colors.greenAccent,
              ),
            ),
          ), // Column
        ),
        FloatingActionButtond(
          _incrementCounter, // `void Function()` => 'onPressed'
          'Increment', // `String` => 'tooltip'
          const Icon(Icons.add), // `Widget` => 'child'
        ),
      ),
    );
  }
}
